### ISSUE TITLE
Build test suite with GitHub Actions

### ISSUE THREAD

**Issue by lpugin** [rism-ch/verovio#1726]
It is now failing on building the doxygen documentation. Error message is
```
The job exceeded the maximum log length, and has been terminated.
```
Weird since it fails after about 5 minutes only. See https://travis-ci.org/github/rism-ch/verovio/builds/733209808

At this stage I would be quite happy to move to GitHub Actions. Help appreciated ;-)

**Comment by musicEnfanthen** [comment]
The only way I see to fix this on Travis would be to make the log output less verbose. At the moment, doxygen build produces more than 28.000 lines of warnings what is terminating the build process. 

Two possible options: 
1) Configure doxygen in a way that it does not show the warnings; 
2) fix the warnings in Verovio code

Additionally, you could try to remove the verbose option in "`travis/ci_deploy-doxygen.sh`",

so `set -ev #` will become `set -e #`,

but my guess is that this is only a drop in the ocean in comparison to the huge amount of warnings.

**Comment by musicEnfanthen** [comment]
Regarding GH actions I would be happy to help. Made a first draft some time ago, maybe it can be useful for further exploration, see https://github.com/musicEnfanthen/verovio/actions

C++ build is running through: https://github.com/musicEnfanthen/verovio/actions/runs/66312718/workflow

JS build is not: https://github.com/musicEnfanthen/verovio/actions/runs/66312719/workflow

**Comment by musicEnfanthen** [comment]
Regarding GH actions I would be happy to help. Made a first draft some time ago, maybe it can be useful for further exploration, see https://github.com/musicEnfanthen/verovio/actions

C++ build is running through: https://github.com/musicEnfanthen/verovio/actions/runs/66312718/workflow

JS build is not: https://github.com/musicEnfanthen/verovio/actions/runs/66312719/workflow

**Comment by lpugin** [comment]
Great, thanks! As a temporary fix I tried to remove the warnings from the doxygen output. Let's see.

Any idea what was not working for the JS build?

**Comment by rettinghaus** [comment]
I also tried a bit with actions, but wasn't too successful so far. 
I thought, this could help with PRs:
https://github.com/HorstBaerbel/action-clang-format

**Comment by musicEnfanthen** [comment]
Seems to work now without the doxygen output?

> Any idea what was not working for the JS build?

It was related to the precaching stage. Did not find emsdk folder with the current setup.

**Comment by musicEnfanthen** [comment]
@rettinghaus Yes, could be good to run clang format checking/linting on every commit.

**Comment by lpugin** [comment]
Additionally, I would like to have the test suite built on PR so we can evaluate the changes. I think this would require a server for the results (images) to be published. This would be the first step.

A second step would be to evaluate where changes occur in comparison with the develop branch. I can see various approaches:
* XML comparison of the SVG tree
* XML comparison of the SVG tree + some attributes (x and y?)
* Pixel-based comparison of images
* XML comparison of the MEI output

These should generate a list of tests where the PR yields changes. This should avoid regressions on unrelated areas to be discovered only later. It would be good to have the report somehow integrated in the GitHub PR so we can straight away see it and look at the cases where something changed.

Does anybody have experience or ideas for putting something like this? (We can have a server up for storing the test results.)

**Comment by musicEnfanthen** [comment]
This will be a huge project, but it would definitely bring great benefits.

To go one step back first: The drafts for the JS and cpp builds are now running successfully. See here: https://github.com/musicEnfanthen/verovio/actions/runs/305304653

However, the JS build triggers so far only one build option (`./buildToolkit -c -H`). From here, it would be pretty easy to setup 4 parallel jobs for all 4 build options. But maybe we can find a solution that does not require to repeat the whole emscripten setup section 4 times?

With Gh actions it is possible to store artifacts of a workflow run, see: https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts . So, the build files can be uploaded as GH artifacts, to be pushed to `gh-pages`.

Ad test-suite: How do you invoke the test suite? (cf. also https://github.com/rism-ch/verovio.org/issues/14)

Ad comparison tools: Some, maybe outdated, starting points are here: https://discourse.opensourcedesign.net/t/existing-tools-attempts-for-diffing-svg/763 . Maybe the music diff tool (https://music-diff.edirom.de/) could also be useful?

**Comment by lpugin** [comment]
Great! We could look how to optimize the 4 builds, but I think the easiest (for now) would be to have them in parallel. Or sequentially using the same setup can work too if this is not too long. (You need to add `-M` to the `./buildToolkit` call in order to regenerate the makefile). Maybe you can try this?

Artifacts looks perfect! Does they have to be push to `gh-pages`?

Invoking the test suite would require to:
* checkout the verovio.org repository
* loop through all files in _tests/
* invoke Verovio for each file

In bash, it would look like something like:
```bash
dir="/path/to/verovio.org/gh-pages/_tests"
outdir="/path/to/output/dir"

cd /path/to/verovio/tools
make -j8

rm $outdir/*.svg;

all=`LS $dir`
for t in $all; do
    for i in `LS $dir/$t/*.mei`; do ./verovio -r ../data/ $i --breaks none --adjust-page-height --scale 50; done
    for i in `LS $dir/$t/*.mei`; do ./verovio -r ../data/ -t mei --remove-ids --all-pages $i; done
    mv $dir/$t/*.svg $outdir;
done
```

Then maybe we need to generate a simple webpage with the SVG files, or something like that, but I guess it depends how artefacts work.

**Comment by musicEnfanthen** [comment]
Thanks a lot, will try this.

The parallel builds and artifact uploads are working now, except for the wasm build. Wasm throws many undefined symbol errors: https://github.com/musicEnfanthen/verovio/runs/1257309439?check_suite_focus=true

Last lines of log:

```
error: undefined symbol: __cxa_pure_virtual (referenced by top-level compiled C/C++ code)
warning: ___cxa_pure_virtual may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: __dynamic_cast (referenced by top-level compiled C/C++ code)
warning: ___dynamic_cast may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/home/runner/work/verovio/verovio/emsdk-cache/emsdk-master/node/12.18.1_64bit/bin/node /home/runner/work/verovio/verovio/emsdk-cache/emsdk-master/upstream/emscripten/src/compiler.js /tmp/tmpac1bbxbf.txt' failed (1)
 Failed.
```

Maybe you have an idea what I am missing here, @lpugin ?

**Comment by lpugin** [comment]
It seems that there is a linking problem with the latest release of the emdk. I asked there https://groups.google.com/g/emscripten-discuss/c/6foEO6Y3etw (Travis is also failing https://travis-ci.org/github/rism-ch/verovio/jobs/735630932)

**Comment by lpugin** [comment]
It should be solved
https://github.com/rism-ch/verovio/commit/9c71733fa7022864ea3f833416a16bf366be870a
https://github.com/emscripten-core/emscripten/commit/e32abfa40ed25006c0e8647e0f302a971f61037b

**Comment by musicEnfanthen** [comment]
Perfect, thanks for the fix. It builds on Travis and GH actions.

Could you please check at the deploy stage in https://github.com/musicEnfanthen/verovio/runs/1258292548?check_suite_focus=true under `git status` if that are the correct files to be pushed later to `gh-pages` and if they are in the correct directories?

**Comment by lpugin** [comment]
Thanks @musicEnfanthen migrating this! For the test suite - I changed the topic of the issue. 

One thing to keep in mind even though it is not a high priority: some test files have specific parameter encoded in the header. For example https://github.com/rism-ch/verovio.org/blob/061f015800e1558ddcebf9e2d330a4927b8e98d0/_tests/score/score-009.mei
```xml
<extMeta><![CDATA[ { "useBraceGlyph": true, "font": "Bravura" }]]></extMeta>
```
In the JS test page, the tag is parsed and passed as JSON to the toolkit. If we want to do the same for the test suite in the actions, which would be good, then it might be easier to use either the JS toolkit or the Python one

**Comment by musicEnfanthen** [comment]
> 
> 
> I also tried a bit with actions, but wasn't too successful so far.
> I thought, this could help with PRs:
> https://github.com/HorstBaerbel/action-clang-format

@rettinghaus How would you invoke clang-format on verovio code? Is there a testfile one could use to test the result of the clang-format operation?

**Comment by musicEnfanthen** [comment]
> 
> 
> ```
> <extMeta><![CDATA[ { "useBraceGlyph": true, "font": "Bravura" }]]></extMeta>
> ```
> 
> In the JS test page, the tag is parsed and passed as JSON to the toolkit. If we want to do the same for the test suite in the actions, which would be good, then it might be easier to use either the JS toolkit or the Python one

Parsing logic is here: https://github.com/rism-ch/verovio.org/blob/gh-pages/_includes/test-version.html#L288-L291

**Comment by musicEnfanthen** [comment]
Ok, tests are basically running via gh-actions: https://github.com/musicEnfanthen/verovio/runs/1272373017?check_suite_focus=true 

But I had to exclude the test categories for `ligature`, `mensural`, `note`, `sameas ` and `stem` . Otherwise gh-actions runner throws a `segmentation fault (code dumped)` error when it tries to call

`./verovio -r ../data/ $i --breaks none --adjust-page-height --scale 50;`

Any ideas what is happening there? I.e., are there any particularities in these categories? Does this also happen at the local level?
But maybe it is not connected to the test cases at all... but to the runner and VE setup, cf. https://github.com/actions/virtual-environments/issues/159

**Comment by lpugin** [comment]
Great! (ligature issue is related to #1752)

**Comment by rettinghaus** [comment]
@musicEnfanthen I was able once to start the action, but it opened a new PR to develop branch. Instead it should add a commit to an open PR. 
Wasn't able to reproduce it. 
See branch https://github.com/rettinghaus/verovio/tree/clang-test for formatting errors.

**Comment by musicEnfanthen** [comment]
Thanks @rettinghaus . Will look into clang-format after fixing #1758 (close to it).

**Comment by musicEnfanthen** [comment]
Ad tests: Meanwhile I was able to isolate the broken tests:

- the complete [ligature folder](https://github.com/rism-ch/verovio.org/tree/gh-pages/_tests/ligature) (46 tests), related to https://github.com/rism-ch/verovio/issues/1726#issuecomment-711788781
- [mensural/mensural-001.mei](https://github.com/rism-ch/verovio.org/blob/gh-pages/_tests/mensural/mensural-001.mei)
- [note/note-003.mei](https://github.com/rism-ch/verovio.org/blob/gh-pages/_tests/note/note-003.mei)
- [sameas/sameas-002.mei](https://github.com/rism-ch/verovio.org/blob/gh-pages/_tests/sameas/sameas-002.mei)
- [stem/stem-002.mei](https://github.com/rism-ch/verovio.org/blob/gh-pages/_tests/stem/stem-002.mei)

Maybe someone is able to figure out if there is something wrong with the last 4 files?

**Comment by lpugin** [comment]
The following are all related to the regression #1726
* mensural/mensural-001.mei
* note/note-003.mei
* stem/stem-002.mei

**Comment by lpugin** [comment]
* sameas/sameas-002.mei related to #1761

**Comment by lpugin** [comment]
See #1762 for discussion on clang-formatting

**Comment by lpugin** [comment]
@musicEnfanthen it seems that gcc-latest on macos uses AppleClang
https://github.com/rism-ch/verovio/runs/1285451643#step:6:23
Is this what to expect?

**Comment by musicEnfanthen** [comment]
@lpugin yes, you are right. Investigating.

**Comment by lpugin** [comment]
> Ad tests: Meanwhile I was able to isolate the broken tests:
> ...
> Maybe someone is able to figure out if there is something wrong with the last 4 files?

@musicEnfanthen all tests should now build successfully

**Comment by musicEnfanthen** [comment]
Great, I'll have a look

**Comment by musicEnfanthen** [comment]
@lpugin Yes, all tests are running now: https://github.com/musicEnfanthen/verovio/runs/1293309020?check_suite_focus=true

**Comment by musicEnfanthen** [comment]
As a next step, we could return the svg graphics as artifacts or upload them to a repo (both approaches are tested in build workflow and should be easy to set up).

I am also investigating how we get the additional parameters from the extMeta tag in gh-actions

**Comment by lpugin** [comment]
I think what we need to have is a webpage with the SVG graphics so we can quickly skim through them. Would this be possible with the artifacts? Or are they only available as zip files? If only a zip is available, then we should upload them to a repository or a server. A repository would be fine, but we don't need to keep the full history. We already have actions that pushes to our own server so this will also be easy to set up. The step would be 
* [ ] run the tests and generate the SVG
* [ ] generate a html page with the SVG (eventually we can add a basic menu to access section, or filter tests that changed when this is in place
* [ ] upload the page and the SVG to the server (or the repo or the artifacts if this is the way to go) within a directory based on PR sha
* [ ] add a link to page in a comment the PR (something like this https://github.com/marketplace/actions/comment-on-pr ?)

@xhero and @ahankinson can see where we can host this.

**Comment by musicEnfanthen** [comment]
Couldn't we reuse the existing comparison website and its logic, or a duplicate of it?

**Comment by ahankinson** [comment]
Is GH Pages an option?

**Comment by lpugin** [comment]
> Edit: Thinking about it, I guess you don't need to store the commit history, so no need to use GH pages. What's the structure of the website? `http://verovio-tests-home.com/tests/[branch-name].html`?

Yes, something like this.

> Would you push just that page? What happens to pages for branches that are deleted / merged?

I think we can keep them for a while and then deleted them. For example, until there is a release. Then we can use the comparison page. The goal is essentially to catch problems during development and avoid regressions as much as possible. 

@musicEnfanthen I would not reuse the comparison page logic, which renders them on the fly. This means having iframes with distinct JS builds, which is cumbersome and can be fairly slow when looking at them. Using the command-line tool with two builds and running all the tests for comparison (develop and PR) seems simpler. Basically
* build the develop version and run all the tests
* build the PR version and run all the tests
* perform some comparison (XML-based, pixel-based) and list the differences
* generate a webpage with both SVG side-by-side (or even superimposed?) and a way to easily see only the ones that changed

**Comment by ahankinson** [comment]
Do you have a preferred hostname? I can set up a new host on our www-production server to host it.

**Comment by lpugin** [comment]
We can go for tests.verovio.org

**Comment by lpugin** [comment]
@musicEnfanthen I added a script call from and action that build the test suite with the Python toolkit and loads the `extMeta` parameters. See:
* https://github.com/lpugin/verovio/blob/develop/.github/workflows/tests_build.yml
* https://github.com/lpugin/verovio/blob/develop/doc/test-suite.py
* https://github.com/lpugin/verovio/runs/1359373604?check_suite_focus=true

I will add one that generates an HTML webpage with the diff. I will then need you help you put things together...

**Comment by musicEnfanthen** [comment]
Fantastic, looks great !

**Comment by lpugin** [comment]
Up and running. Every PR will now be evaluated against the develop branch and list all tests with an image diff. See
https://github.com/rism-ch/verovio/pull/1810#issuecomment-724272426 and https://test-suite.verovio.org/cdb7266adc1756082e4f128b90241870410f338b/
I can certainly be improved but this is a good start.

**Comment by musicEnfanthen** [comment]
Congratulations, this is brilliant.
